### PR TITLE
JC-46. Join participant to competition

### PR DESCRIPTION
**Pull request by saffchen**

### CODE REVIEW

File: src/main/java/org/jacp/controller/CompetitionController.java [hunk 1]
```diff
@@ -77,6 +77,16 @@ public ResponseEntity<CompetitionDto> startCompetition(@PathVariable Long compet
         return ResponseEntity.ok(competitionDto);
     }
 
+    @PostMapping("/{competitionId}/join")
+    public ResponseEntity<CompetitionDto> joinParticipants(@PathVariable Long competitionId,
```
Comment (pelmegov) [inline src/main/java/org/jacp/controller/CompetitionController.java]: `joinParticipant` or just `join` for the method name will be more clear.

File: src/main/java/org/jacp/controller/CompetitionController.java [hunk 2]
```diff
@@ -77,6 +77,16 @@ public ResponseEntity<CompetitionDto> startCompetition(@PathVariable Long compet
...
+                                                           @RequestParam Long participantId) {
+        CompetitionDto competitionDto =
+                mapper.toCompetitionEntityByCompetitionDto(competitionService.getCompetitionEntity(competitionId));
+        competitionDto.joinParticipant(participantId);
+        competitionService.create(mapper.toCompetitionEntity(competitionDto));
```
Comment (pelmegov) [inline src/main/java/org/jacp/controller/CompetitionController.java]: It's not a common practice to transfer entity object to dto and after some changes move it back.
It's much easier to just have a one more service method: `competitionService.joinParticipant(competitionId, participantId) ` and inside the method just do all required changes on the entity.

Comment (saffchen) [reply src/main/java/org/jacp/controller/CompetitionController.java]: Fixed

File: src/main/java/org/jacp/controller/CompetitionController.java (hunk 1 above)
Comment (saffchen) [reply src/main/java/org/jacp/controller/CompetitionController.java]: Fixed

File: src/main/java/org/jacp/controller/CompetitionController.java [hunk 3]
```diff
@@ -67,14 +65,14 @@ public ResponseEntity<CompetitionEntity> createCompetition(@RequestBody SearchDt
...
-    public ResponseEntity<CompetitionDto> startCompetition(@PathVariable Long competitionId) {
-        CompetitionDto competitionDto =
-                mapper.toCompetitionEntityByCompetitionDto(competitionService.getCompetitionEntity(competitionId));
-        competitionDto.setStartDate(Date.from(Instant.now()));
-        competitionDto.calculateEndDate();
-        competitionDto.setStatus(Status.QUEUED.toString());
-        competitionService.create(mapper.toCompetitionEntity(competitionDto));
-        return ResponseEntity.ok(competitionDto);
+    public ResponseEntity<CompetitionEntity> startCompetition(@PathVariable Long competitionId) {
+        return ResponseEntity.ok(competitionService.startCompetition(competitionId));
```
Comment (pelmegov) [inline src/main/java/org/jacp/controller/CompetitionController.java]: It's better to map response to DTO.

File: src/main/java/org/jacp/service/CompetitionService.java [hunk 4]
```diff
@@ -27,7 +29,23 @@ public List<CompetitionEntity> getAllByStatus(Status status) {
...
+    public CompetitionEntity getCompetitionEntity(Long id) {
         return competitionRepository.getCompetitionEntityById(id);
     }
+
+    public CompetitionEntity startCompetition(Long id) {
+        CompetitionEntity competitionEntity = getCompetitionEntity(id);
+        competitionEntity.setStartDate(Date.from(Instant.now()));
+        competitionEntity.calculateEndDate();
+        competitionEntity.setStatus(Status.QUEUED);
+        create(competitionEntity);
```
Comment (pelmegov) [inline src/main/java/org/jacp/service/CompetitionService.java]: You don't need to call save() method here. Just annotate the method as transactional.

File: src/main/java/org/jacp/controller/CompetitionController.java (hunk 3 above)
Comment (saffchen) [reply src/main/java/org/jacp/controller/CompetitionController.java]: Fixed

File: src/main/java/org/jacp/service/CompetitionService.java (hunk 4 above)
Comment (saffchen) [reply src/main/java/org/jacp/service/CompetitionService.java]: Fixed

**Review by pelmegov — APPROVED** [review]

**Pull request merged**